src/Microsoft.AspNet.Mvc.TagHelpers/LinkTagHelper.cs

```diff
@@ -307,45 +310,66 @@ private void BuildGlobbedLinkTags(TagHelperAttributeList attributes, TagHelperCo
+            // indicating that the primary stylesheet failed to load.
+            // GetEmbeddedJavaScript returns JavaScript to which we add '"{0}","{1}",{2});'
+            builder
+                .AppendHtml("<script>")
+                .AppendHtml(JavaScriptResources.GetEmbeddedJavaScript(FallbackJavaScriptResourceName))
+                .AppendHtml("\"")
+                .AppendHtml(JavaScriptEncoder.Encode(FallbackTestProperty))
+                .AppendHtml("\",\"")
+                .AppendHtml(JavaScriptEncoder.Encode(FallbackTestValue))
+                .AppendHtml("\",");
+
+            builder.AppendHtml("[");
+
+                // Perf: Avoid allocating enumerator
+            var firstAdded = false;
+            for (var i = 0; i < fallbackHrefs.Count; i++)
+            {
+                if (firstAdded)
+                {
+                    builder.AppendHtml(",\"");
+                }
+                else
+                {
+                    builder.AppendHtml("\"");
+                    firstAdded = true;
                 }
 
-                builder.AppendHtml(HtmlString.NewLine);
+                // fallbackHrefs come from bound attributes and globbing. Must always be non-null.
+                Debug.Assert(fallbackHrefs[i] != null);
+                var valueToWrite = fallbackHrefs[i];
+                if (AppendVersion == true)
+                {
+                    valueToWrite = _fileVersionProvider.AddFileVersionToPath(fallbackHrefs[i]);
+                }
 
-                // Build the <meta /> tag that's used to test for the presence of the stylesheet
-                builder
-                    .AppendHtml("<meta name=\"x-stylesheet-fallback-test\" content=\"\" class=\"")
-                    .Append(FallbackTestClass)
-                    .AppendHtml("\" />");
-
-                // Build the <script /> tag that checks the effective style of <meta /> tag above and renders the extra
-                // <link /> tag to load the fallback stylesheet if the test CSS property value is found to be false,
-                // indicating that the primary stylesheet failed to load.
-                // GetEmbeddedJavaScript returns JavaScript to which we add '"{0}","{1}",{2});'
-                builder
-                    .AppendHtml("<script>")
-                    .AppendHtml(JavaScriptResources.GetEmbeddedJavaScript(FallbackJavaScriptResourceName))
-                    .AppendHtml("\"")
-                    .AppendHtml(JavaScriptEncoder.Encode(FallbackTestProperty))
-                    .AppendHtml("\",\"")
-                    .AppendHtml(JavaScriptEncoder.Encode(FallbackTestValue))
-                    .AppendHtml("\",")
-                    .AppendHtml(JavaScriptStringArrayEncoder.Encode(JavaScriptEncoder, fallbackHrefs))
-                    .AppendHtml(");</script>");
+                builder.AppendHtml(JavaScriptEncoder.Encode(valueToWrite));
```

Comment:
I tried using the `HtmlContentTextWriter` here - however the encoder writes values as characters and we end up allocating strings for each of these.